pkg/exporter/process.go

```diff
@@ -101,55 +98,44 @@ func InitExportingProcess(collectorAddr net.Addr, obsID uint32, tempRefTimeout u
 	return expProc, nil
 }
 
-func (ep *ExportingProcess) AddRecordAndSendMsg(recType entities.ContentType, rec entities.Record) (int, error) {
-	if recType == entities.Template {
-		ep.updateTemplate(rec.GetTemplateID(), rec.GetTemplateElements(), rec.GetMinDataRecordLen())
-	} else if recType == entities.Data {
-		err := ep.dataRecSanityCheck(rec)
-		if err != nil {
-			return 0, fmt.Errorf("AddRecordAndSendMsg: error when doing sanity check:%v", err)
+func (ep *ExportingProcess) AddSetAndSendMsg(setType entities.ContentType, set entities.Set) (int, error) {
+	recBytes := make([]byte, 0)
+	for _, record := range set.GetRecords() {
+		if setType == entities.Template {
+			ep.updateTemplate(record.GetTemplateID(), record.GetInfoElements(), record.GetMinDataRecordLen())
+		} else if setType == entities.Data {
+			err := ep.dataRecSanityCheck(record)
+			if err != nil {
+				return 0, fmt.Errorf("AddRecordAndSendMsg: error when doing sanity check:%v", err)
+			}
+		}
+		for _, byte := range record.GetBuffer().Bytes() {
+			recBytes = append(recBytes, byte)
 		}
 	}
-	recBytes := rec.GetBuffer().Bytes()
 
 	msgBuffer := ep.msg.GetMsgBuffer()
 	var bytesSent int
-	// Check if message is exceeding the limit with new record
-	if uint16(msgBuffer.Len()+len(recBytes)) > entities.MaxTcpSocketMsgSize {
-		ep.set.FinishSet()
-		b, err := ep.sendMsg()
-		if err != nil {
-			return b, err
-		}
-		bytesSent = bytesSent + b
+	// Check if message is exceeding the limit with new set
+	if uint16(msgBuffer.Len()+int(set.GetBuffLen())) > entities.MaxTcpSocketMsgSize {
+		return bytesSent, fmt.Errorf("AddSetAndSendMsg: set size exceeds max socket size.")
 	}
 	if msgBuffer.Len() == 0 {
 		err := ep.createNewMsg()
 		if err != nil {
-			return bytesSent, fmt.Errorf("AddRecordAndSendMsg: error when creating message: %v", err)
+			return bytesSent, fmt.Errorf("AddSetAndSendMsg: error when creating message: %v", err)
 		}
 	}
-	// Check set buffer length and type change to create new set in the message
-	if ep.set.GetBuffLen() == 0 {
-		ep.set.CreateNewSet(recType, rec.GetTemplateID())
-	} else if ep.set.GetSetType() != recType {
-		ep.set.FinishSet()
-		ep.set.CreateNewSet(recType, rec.GetTemplateID())
-	}
-	// Write the record to the set
-	err := ep.set.WriteRecordToSet(&recBytes)
-	if err != nil {
-		return bytesSent, fmt.Errorf("AddRecordAndSendMsg: %v", err)
-	}
-	if recType == entities.Data && !ep.msg.GetDataRecFlag() {
+
+	if setType == entities.Data && !ep.msg.GetDataRecFlag() {
 		ep.msg.SetDataRecFlag(true)
 	}
 
 	// Send the message right after attaching the record
```

Comment:
Remove the comment. May be adding a comment why finish set is needed is better here.